compiler/src/org.graalvm.compiler.nodes/src/org/graalvm/compiler/nodes/java/RegisterFinalizerNode.java

```diff
@@ -101,13 +101,15 @@ public void generate(NodeLIRBuilderTool gen) {
     public static boolean mayHaveFinalizer(ValueNode object, MetaAccessProvider metaAccess, Assumptions assumptions) {
         ObjectStamp objectStamp = (ObjectStamp) object.stamp(NodeView.DEFAULT);
         ResolvedJavaType objectType = StampTool.typeOrNull(objectStamp, metaAccess);
-        if (objectStamp.isExactType()) {
-            return objectType.hasFinalizer();
-        }
-        AssumptionResult<Boolean> result = objectType.hasFinalizableSubclass();
-        if (result.canRecordTo(assumptions)) {
-            result.recordTo(assumptions);
-            return result.getResult();
+        if (objectType != null) {
+            if (objectStamp.isExactType()) {
+                return objectType.hasFinalizer();
+            }
+            AssumptionResult<Boolean> result = objectType.hasFinalizableSubclass();
+            if (result.canRecordTo(assumptions)) {
+                result.recordTo(assumptions);
+                return result.getResult();
+            }
         }
         return true;
```

Comment:
For native image, we need to make sure that `mayHaveFinalizer` always returns `false` because finalizer are not supported at all. This is currently the case because `objectType.hasFinalizableSubclass` always returns `false`. But now that code is no longer called when `objectType` is `null`. So when `objectType == null`, you actually need to replace it with `java.lang.Object` and still do the check for `hasFinalizableSubclass`.

Reply:
OK, so essentially we need to bring back the functionality from https://github.com/oracle/graal/commit/0265c27a2c3d4711bb31f89c124f46a80f189f58

```
        if (objectType == null) {
            objectType = metaAccess.lookupJavaType(Object.class);
        }
```

Reply:
Had to swap in all my knowledge from June when that was done...

Actually, the `StampTool.typeOrNull(objectStamp, metaAccess);` already does the lookup of `java.lang.Object`:
```
    public static ResolvedJavaType typeOrNull(Stamp stamp, MetaAccessProvider metaAccess) {
        if (stamp instanceof AbstractObjectStamp && stamp.hasValues()) {
            AbstractObjectStamp abstractObjectStamp = (AbstractObjectStamp) stamp;
            ResolvedJavaType type = abstractObjectStamp.type();
            if (type == null) {
                return metaAccess.lookupJavaType(Object.class);
            } else {
                return type;
            }
        }
        return null;
    }
```

That means in your case you are either seeing a stamp that is not a `AbstractObjectStamp` (which means something is seriously wrong), or you are seeing a stamp where `hasValues()` returns `false` (which also means something is seriously wrong).

Thanks to your reproducer, I quickly found out what is going on: We are compiling the constructor of `org.sqlite.BusyHandler`. This is an abstract class, and the static analysis found no subclasses. So why is the constructor compiled? Because it is registered as a JNI entry point.

So the real bug is that our JNI stub for constructors does not check if the class is abstract. We end up with an abstract class that is also marked as instantiated - that can lead to all sorts of follow-up problems. Just avoiding the NullPointerException probably makes that particular build go through, but there could still be all sorts of strange other problems then, so even as a workaround this is not a feasible solution.

I will think about a proper fix for this problem, with some added guarantees that an abstract class is never marked as instantiated.

Reply:
Good catch! Although I checked the source code of sqlite-jdbc apparently I didn't pay enough attention to notice that the classes causing the issue are abstract. It's still strange how this only pops up with sqlite :/

Thanks for looking into it and for the explanation.